dashboard/app/models/user.rb

```diff
@@ -1138,6 +1138,67 @@ def has_activity?
     user_levels.attempted.exists?
   end
 
+  # Returns the next visible script_level for the next progression level in the # given script that hasn't yet been passed, starting at the last level the
+  # the user most recently submitted
+  def next_unpassed_visible_progression_level(script)
+    # If all levels in the script are complete, no need to find the next level,
+    # will be redirected to /congrats.
+    unless completed?(script)
+      visible_sls = visible_script_levels(script).reject {|sl| sl.bonus || sl.level.unplugged? || sl.locked?(self)}
+
+      sl_level_ids = visible_sls.map(&:level_ids).flatten
+
+      # Levels the user made progress in
+      ul_level_ids = user_levels_by_level(script).keys
+
+      visible_completed_level_ids = sl_level_ids & ul_level_ids
+      visible_incomplete_level_ids = sl_level_ids - ul_level_ids
+
+      first_visible_level = visible_sls.min_by(&:chapter)
+
+      completed_all_visible_levels = visible_incomplete_level_ids.empty?
+
+      # The user has not made any progress or has completed all visible levels
+      # but not the entire script, return the first visible script_level
+      if ul_level_ids.empty? || completed_all_visible_levels
+        return first_visible_level
+      end
```

Comment:
nit: this (and the `most_recent_sl == last_visible_level` check below) can just be `return first_visible_level if ul_level_ids.empty? || completed_all_visible_levels`